src/lightmap.cpp

```diff
@@ -1249,46 +1250,124 @@ void map::apply_light_arc( const tripoint &p, const units::angle &angle, float l
 
     apply_light_source( p, LIGHT_SOURCE_LOCAL );
 
+    const point p2( p.xy() );
+
+    auto &cache = get_cache( p.z );
+    four_quadrants( &lm )[MAPSIZE_X][MAPSIZE_Y] = cache.lm;
+    float( &transparency_cache )[MAPSIZE_X][MAPSIZE_Y] = cache.transparency_cache;
+
     // Normalize (should work with negative values too)
-    const units::angle wangle = wideangle / 2.0;
+    units::angle wangle = wideangle / 2.0;
 
-    units::angle nangle = fmod( angle, 360_degrees );
+    units::angle nangle = normalize( angle );
+    units::angle oangle = angle - wangle;
+    units::angle cangle = angle + wangle;
 
     tripoint end;
     int range = LIGHT_RANGE( luminance );
     calc_ray_end( nangle, range, p, end );
-    apply_light_ray( lit, p, end, luminance );
 
     tripoint test;
     calc_ray_end( wangle + nangle, range, p, test );
 
     const float wdist = hypot( end.x - test.x, end.y - test.y );
     if( wdist <= 0.5 ) {
+        apply_light_ray( lit, p, end, luminance );
         return;
     }
 
     // attempt to determine beam intensity required to cover all squares
     const units::angle wstep = ( wangle / ( wdist * M_SQRT2 ) );
 
+    if( wideangle >= 45_degrees ) {
+        //cut pre-subsection
+        if( fmod( oangle, 45_degrees ) != 0_degrees ) {
+            units::angle preangle = oangle;
+            oangle = 45_degrees * std::ceil( to_degrees( oangle ) / 45 );
+            units::angle preanglew = units::from_radians( std::abs( ( oangle - preangle ).value() ) );
+
+            // NOLINTNEXTLINE(clang-analyzer-security.FloatLoopCounter)
+            for( units::angle ao = 0_degrees; ao <= preanglew; ao += wstep ) {
+                if( trigdist ) {
+                    double fdist = ( ao * M_PI_2 ) / preanglew;
+                    end.x = static_cast<int>(
+                                p.x + ( static_cast<double>( range ) - fdist * 2.0 ) * cos( preangle + ao ) );
+                    end.y = static_cast<int>(
+                                p.y + ( static_cast<double>( range ) - fdist * 2.0 ) * sin( preangle + ao ) );
+                    apply_light_ray( lit, p, end, luminance );
+                } else {
+                    calc_ray_end( nangle + ao, range, p, end );
+                    apply_light_ray( lit, p, end, luminance );
+                }
+            }
+        }
+        int numoct = to_degrees( cangle - oangle ) / 45;
+        int firstoct = to_degrees( oangle ) / 45;
```

Comment:
To get rid of type casting warnings
```suggestion
        int numoct = std::floor( to_degrees( cangle - oangle ) / 45 );
        int firstoct = static_cast<int>( std::lround( to_degrees( oangle ) / 45 ) );
```